enhancements/network/openstack-dual-stack.md

```diff
@@ -0,0 +1,188 @@
+additional addresses, like storage interfaces.
+- Pods are configured with both IPv4 and IPv6 addresses.
+
+### Non-Goals
+
+- Support for OpenStack endpoints configured with dual-stack.
+- Support for LoadBalancer type Services.
+- Installer (IPI) handles creation of the Network and dual-stack Subnet.
+
+## Proposal
+
+Customers will be able to specify the IPv4 and IPv6 Subnets
+meant to be used in the `install-config.yaml`.
+
+### Workflow Description
+
+1. The customers will configure a network (tenant or provider) with IPv4 and
+IPv6 subnets. The IPv6 address mode for the Subnet can be of any type:
+SLAAC, Stateless or Stateful. Here is a sample of network and subnets creation:
+
+```bash
+$ openstack network create dual-stack
+$ openstack subnet create --network dual-stack --subnet-range 10.0.0.0/16 ipv4-subnet
+$ openstack subnet create --network dual-stack --ipv6-address-mode slaac --ipv6-ra-mode slaac --subnet-range 2001:db8:2222:5555::/64 --ip-version 6 ipv6-subnet
+```
+
+2. The customer will create API and Ingress Ports using the network created in the
+previous step. Here is a sample of ports creation:
+
+```bash
+$ openstack port create api --network dual-stack
+$ openstack port create ingress --network dual-stack
+```
+
+3. The customer will write the `install-config.yaml` and add the subnet IDs to
+the new field `machinesSubnets`.
+4. The customer will generate the manifests and configured MachineConfigs
+for the master and workers depending on the IPv6 addresses mode chosen for the Subnets
+in step 1. If SLAAC or Stateless was used the MachineConfigs has to contain `ip_addr_mode=eui64`
+and `method=auto` in the connection for the ethernet interface. While when using stateful
+it has to enable `ip=dhcp,dhcpv6` in the kernel arguments.
+5. The `uuid` from the network has to be added in the worker MachineSet alongside with the
+subnets that were pre-filled for the machine-api-provider-openstack to create one unique Port
+with both addresses configured for the Server.
+6. The CIDRs of the subnets have to added to the `cloud-provider-config.yaml`
+in the `address-sort-order` field to make sure IPv4 address will have higher priority in the
+Node addresses.
+7. The installer will validate the subnets and VIPs, add the security group to the api and ingress
+ports and link the floating IPs to those Ports. Also, create the additional IPv6 security group rules
+and configured the servers with dual-stack addresses and add the dual-stack allowed address pairs.
+8. In machine-config-operator, kubelet will be update to not be configured with node-ips.
+9. cloud-provider-openStack will filter the node addresses that are meant to be specified in the Status of the Node
+by looking for a tag that was applied in the main Ports of the Node.
+
+### API Extensions
+
+The `machinesSubnet` field, which is available in the `install-config.yaml` under
+`platform` and `openstack` will be deprecated in favor of a new field named `machinesSubnets`
```

Comment:
As an alternative to `machinesSubnets` you could consider accepting a structure like [CAPO's `portOpts`](https://github.com/openshift/cluster-api-provider-openstack/blob/8bd9c35fa7dfa81697f9e62b9ea598cba699ca02/pkg/apis/openstackproviderconfig/v1alpha1/types.go#L202). With that, you can define a port in a flexible way, and it can attach to as many subnets as you wish (and if you want the user to be able to define multiple ports, you can accept an array of `portOpts`, if we can't assume that both the IPv4 and the IPv6 subnets will exist on the same network).

We are going to use that same `portOpts` in failure domains, and we could generalise its use in the Installer API.

Reply:
done